README.md
# HiPay Professional for GoodLayers Tourmaster integration

[![GitHub license](https://img.shields.io/badge/license-Apache%202-blue.svg)](https://raw.githubusercontent.com/hipay/hipay-enterprise-sdk-prestashop/master/LICENSE.md)

## Getting started

To use **Hipay Professional** payments you need to previously create an account. To create a sandbox (test purposes) account please use https://test-www.hipaydirect.com/registration/register and for a live account please use https://www.hipaydirect.com/registration/register.

## Configuration

**What do you need to start?**
- Get your Hipay account webservice access - login and password - from Hipay Professional Administration;
- Create a website, get the ID of the website and the id of the category;

**Get the website category id**
Replace WEBSITEID in one of the folllowing urls, depending on the type of your account: Live or Sandbox (test). For a sandbox account use https://test-payment.hipay.com/order/list-categories/id/WEBSITEID and for a live account use https://payment.hipay.com/order/list-categories/id/WEBSITEID. Use one of the IDs of the list.

**Set your Hipay account in Tourmaster plugin**
- Hipay Live Mode: on for a live account and off for a sandbox account;
- Hipay Webservice Login: webservice access login from your Hipay account;
- Hipay Webservice Password: webservice access password from your Hipay account;
- Hipay Website Category: one of the categories ids of the website;
- Hipay Website Shop ID: id of the shop associated with the website (if you have created one);
- Logo URL: absolute link to an image that is shown in Hipay payment window - usually the website logo;
- Technical Email: email address that receives the result of payment notifications;
- Hipay Website Rating: choose between: ALL, +18, +16 or +12;
- Hipay Account Currency: 3 digit code of the currency of the account;
- Minimum amount: minimum amount to use Hipay as payment method (not yet used);
- Maximum amount: maximum amount to use Hipay as payment method (not yet used);

## Requirements

Uses **SOAP Extension** to generate payments and **SimpleXML Extension** to process payments notifications. Configuration panel will let you know if these extensions are active or not.

## Features

Provide local and international payment methods using an Hipay Professional account, integrated with **GoodLayers Tourmaster**. **GoodLayers Tourmaster**, a Premium WordPress plugin that comes with most complete tour management system. This plugin is especially designed for tour operator, tour package, travel agency, travel ticket seller, transportation ticket and much more.

## License

The **HiPay Professional for GoodLayers Tourmaster integration** is available under the **Apache 2.0 License**. Check out the [license file][project-license] for more information.

[project-issues]: https://github.com/hipaypt/hipay-professional-tourmaster/issues
[project-license]: LICENSE.md
[project-changelog]: CHANGELOG.md


CHANGELOG.md
# Version 1.0.0

- Configure one currency account
- Payment support for automatic capture
- Payment notification processing


includes/hipayprofessional.php
<?php
/* 
Version: 1.0.0
Author: Diogo Ferreira - HiPay Portugal
License: Apache License 2.0
*/

/*
* CONFIGURATIONS
*/

add_filter('goodlayers_plugin_payment_option', 'tourmaster_hipayprofessional_payment_option');
if( !function_exists('tourmaster_hipayprofessional_payment_option') ){
	function tourmaster_hipayprofessional_payment_option( $options ){

		if (extension_loaded('soap')) {
			$lib_title 	= esc_html__('SOAP Extension active.','tourmaster') . "<br>";
		} else {
			$lib_title 	= esc_html__('Please Install / Activate SOAP Extension.','tourmaster') . "<br>";
		}
		if (extension_loaded('simplexml')) {
			$lib_title .= esc_html__('SimpleXML Extension active.','tourmaster');
		} else {
			$lib_title .= esc_html__('Please Install / Activate SimpleXML Extension.','tourmaster');
		}

		$options['hipayprofessional'] = array(
			'title' => esc_html__('Hipay Professional', 'tourmaster'),
			'options' => array(
				'hipayprofessional-live-mode' => array(
					'title'	 		=> __('Hipay Live Mode', 'tourmaster'),
					'type' 			=> 'checkbox',
					'default' 		=> 'disable',
					'description' 	=> esc_html__('Disable this option to test on sandbox mode.', 'tourmaster')
				),
				'hipayprofessional-merchant-login' => array(
					'title' 		=> esc_html__('Hipay Webservice Login', 'tourmaster'),
					'type' 			=> 'text'
				),
				'hipayprofessional-merchant-password' => array(
					'title' 		=> esc_html__('Hipay Webservice Password', 'tourmaster'),
					'type' 			=> 'text',	
				),
				'hipayprofessional-website' => array(
					'title' 		=> esc_html__('Hipay Website ID', 'tourmaster'),
					'type' 			=> 'text',	
				),
				'hipayprofessional-website-category' => array(
					'title' 		=> esc_html__('Hipay Website Category', 'tourmaster'),
					'type' 			=> 'text',	
					'description' 	=> esc_html__('Replace WEBSITE ID in one of the following URLs, according the platform you are using, choosing one of the categories ID.', 'tourmaster') . '<br>' . esc_html__('Live Platform: ', 'tourmaster') . '<br>' . esc_html__('Sandbox Platform: ', 'tourmaster')
				),
				'hipayprofessional-website-shopid' => array(
					'title' 		=> esc_html__('Hipay Website Shop ID', 'tourmaster'),
					'type' 			=> 'text',	
					'description' 	=> esc_html__('If you have a shop associated with your Website, please provide the Shop ID.', 'tourmaster') 
				),
				'hipayprofessional-website-logo' => array(
					'title' 		=> esc_html__('Logo URL', 'tourmaster'),
					'type' 			=> 'text',	
					'description' 	=> esc_html__('Absolute link to an image that is shown in Hipay payment window.', 'tourmaster')
				),
				'hipayprofessional-email-notification' => array(
					'title' 		=> esc_html__('Technical Email', 'tourmaster'),
					'type' 			=> 'text',	
					'description' 	=> esc_html__('Receive the result of callback notifications on this email address.', 'tourmaster')
				),
				'hipayprofessional-website-rating' => array(
					'title' 		=> esc_html__('Hipay Website Rating', 'tourmaster'),
					'type' 			=> 'text',	
					'default' 		=> 'ALL',
					'description' 	=> esc_html__('Choose between: ALL, +18, +16 or +12.', 'tourmaster')
				),
				'hipayprofessional-currency-code' => array(
					'title' 		=> esc_html__('Hipay Account Currency', 'tourmaster'),
					'type' 			=> 'text',	
					'default' 		=> 'EUR',
					'description'	=> esc_html__('3 digit currency code. Example: EUR, USD, etc.', 'tourmaster')
				),
				'hipayprofessional-min-amount' => array(
					'title' 		=> esc_html__('Minimum amount', 'tourmaster'),
					'type' 			=> 'text',	
					'default' 		=> '2',
					'description' 	=> esc_html__('Minimum amount to use Hipay as payment method.', 'tourmaster')
				),
				'hipayprofessional-max-amount' => array(
					'title' 		=> esc_html__('Maximum amount', 'tourmaster'),
					'type' 			=> 'text',	
					'default' 		=> '2500',
					'description' 	=> esc_html__('Maximum amount to use Hipay as payment method.', 'tourmaster')
				),
				'hipayprofessional-libs' => array(
					'title' 		=> $lib_title,
					'type' 			=> 'select',
					'description' 	=> esc_html__('Hipay uses SOAP to generate payments and SimpleXML to process payments notifications.','tourmaster')
				),
			)
		);
		return $options;
	} 
}


/*
* PAYMENT WINDOW REDIRECTION
*/

add_filter('goodlayers_hipayprofessional_payment_form', 'tourmaster_hipayprofessional_payment_url', 10, 2);
if( !function_exists('tourmaster_hipayprofessional_payment_url') ){
	function tourmaster_hipayprofessional_payment_url( $ret = '', $tid = '' ){
		
		$live_mode 			= tourmaster_get_option('payment', 'hipayprofessional-live-mode', 'disable');
		$username 			= tourmaster_get_option('payment', 'hipayprofessional-merchant-login', '');
		$password 			= tourmaster_get_option('payment', 'hipayprofessional-merchant-password', '');
		$website 			= tourmaster_get_option('payment', 'hipayprofessional-website', '');
		$website_category 	= tourmaster_get_option('payment', 'hipayprofessional-website-category', '');
		$website_rating 	= tourmaster_get_option('payment', 'hipayprofessional-website-rating', '');
		$website_currency 	= tourmaster_get_option('payment', 'hipayprofessional-currency-code', '');
		$website_logo 		= tourmaster_get_option('payment', 'hipayprofessional-website_logo', '');
		$website_shop 		= tourmaster_get_option('payment', 'hipayprofessional-website-shopid', '');
		$technical_email	= tourmaster_get_option('payment', 'hipayprofessional-email-notification', '');

		$tourmaster_data 	= apply_filters('goodlayers_payment_get_transaction_data', array(), $tid, array('price', 'tour_id','email'));
		$amount 			= floatval($tourmaster_data['price']);
		$customerEmail		= $tourmaster_data['email'];
		$ws_url 			= ($live_mode != "disable") ? "https://ws.hipay.com/soap/payment-v2/generate?wsdl" : "https://test-ws.hipay.com/soap/payment-v2/generate?wsdl";
		$language 			= get_locale();
		$ip 				= $_SERVER['REMOTE_ADDR'];
		$currentDate 		= date('Y-m-dTH:i:s');

		try
		{
			$client = new SoapClient($ws_url);
			$parameters = new stdClass(); 
			$parameters->parameters = array(
				'wsLogin' 			=> $username,
				'wsPassword' 		=> $password,
				'websiteId' 		=> $website,
				'categoryId' 		=> $website_category,
				'currency' 			=> $website_currency,
				'amount' 			=> $amount,
				'rating' 			=> $website_rating,
				'locale' 			=> $language,
				'customerIpAddress' => $ip,
				'merchantReference' => $tid,
				'description' 		=> "#" . $tid,
				'executionDate' 	=> $currentDate,
				'manualCapture' 	=> 0,
				'customerEmail' 	=> $customerEmail,
				'merchantComment' 	=> '',
				'emailCallback' 	=> $technical_email,
				'urlCallback' 		=> add_query_arg(array('tourmaster_hipayprofessional'=>$tid), home_url('/')),
				'urlAccept' 		=> add_query_arg(array('tid' => $tid, 'step' => 4, 'payment_method' => 'hipayprofessional'), tourmaster_get_template_url('payment')),
				'urlDecline' 		=> add_query_arg(array(), tourmaster_get_template_url('payment')),
				'urlCancel' 		=> add_query_arg(array(), tourmaster_get_template_url('payment')),
				'urlLogo' 			=> $website_logo,
			);					

			if ($website_shop != "") $parameters->parameters["shopId"] = $website_shop;
			$result = $client->generate($parameters);

		}
		catch (Exception $e){
			$result->generateResult->code 			= -1;
			$result->generateResult->description 	= $e->getMessage();
		}
		
		ob_start();
		if ($result->generateResult->code == 0) {
		?>
			<input type="hidden" value="<?php echo $result->generateResult->redirectUrl;?>" id="hipayprofessional_url" name="hipayprofessional_url" >
			<div class="goodlayers-hipayprofessional-redirecting-message" ><?php esc_html_e('Please wait while we redirect you to Hipay payment page.', 'tourmaster') ?></div>
			<script type="text/javascript">
				(function($){
					document.location = $("#hipayprofessional_url").val();
				})(jQuery);
			</script>

		<?php
		} else {
		?>
			<div class="goodlayers-hipayprofessional-redirecting-message" ><?php esc_html_e('There was an error generating the payment. Please refresh the page and try again.', 'tourmaster') ?></div>
		<?php				
		}
		$ret = ob_get_contents();
		ob_end_clean();

		return $ret;

	} 
}

/*
* FILTERS
*/
add_filter('tourmaster_hipayprofessional_button_atts', 'tourmaster_hipayprofessional_button_attribute');
if( !function_exists('tourmaster_hipayprofessional_button_attribute') ){
	function tourmaster_hipayprofessional_button_attribute( $attributes ){
		$min_amount	= tourmaster_get_option('payment', 'hipayprofessional-min-amount', '1');
		$max_amount	= tourmaster_get_option('payment', 'hipayprofessional-max-amount', '2500');
		return array('method' => 'ajax', 'type' => 'hipayprofessional', 'min_amount' => $min_amount, 'max_amount' => $max_amount);
	}
}

/*
* CALLBACK NOTIFICATION PROCESSING
*/
add_action('init', 'tourmaster_hipayprofessional_process_callback');
if( !function_exists('tourmaster_hipayprofessional_process_callback') ){
	function tourmaster_hipayprofessional_process_callback(){

		if( isset($_GET['tourmaster_hipayprofessional']) ){

			$payment_info = array(
				'payment_method' => 'hipayprofessional'
			);

			$live_mode = tourmaster_get_option('payment', 'hipayprofessional-live-mode', '');
			if( empty($live_mode) || $live_mode == 'disable' ){
				$hipayprofessional_action_url = "https://test-ws.hipay.com/soap/transaction-v2?wsdl";
			}else{
				$hipayprofessional_action_url = "https://ws.hipay.com/soap/transaction-v2?wsdl";
			}

				$xml = $_POST['xml'];

				$operation = '';
				$status = '';
				$date = '';
				$time = '';
				$transid = '';
				$origAmount = '';
				$origCurrency = '';
				$idformerchant = '';
				$merchantdatas = array();
				$ispayment = true;

				$xml = trim($xml);
				$xml_count = strpos($xml,"<mapi>");
				$xml_len = strlen($xml);
				$xml = substr($xml,$xml_count,$xml_len - $xml_count);

				$obj = new SimpleXMLElement($xml);
				if (isset($obj->result[0]->operation))
					$operation=$obj->result[0]->operation;
				else
					$ispayment =  false;

				if (isset($obj->result[0]->status))
					$status=$obj->result[0]->status;
				else 
					$ispayment =  false;

				if (isset($obj->result[0]->date))
					$date=$obj->result[0]->date;
				else 
					$ispayment =  false;

				if (isset($obj->result[0]->time))
					$time=$obj->result[0]->time;
				else 
					$ispayment =  false;

				if (isset($obj->result[0]->transid))
					$transid=(string)$obj->result[0]->transid;
				else 
					$ispayment =  false;

				if (isset($obj->result[0]->origAmount))
					$origAmount=(string)$obj->result[0]->origAmount;
				else 
					$ispayment =  false;

				if (isset($obj->result[0]->origCurrency))
					$origCurrency=(string)$obj->result[0]->origCurrency;
				else 
					$ispayment = false;

				if (isset($obj->result[0]->idForMerchant))
					$idformerchant=$obj->result[0]->idForMerchant;
				else 
					$ispayment =  false;


				if ($status=="ok" && $operation=="capture") {

					$client 	= new SoapClient($hipayprofessional_action_url);
					$username 	= tourmaster_get_option('payment', 'hipayprofessional-merchant-login', '');
					$password 	= tourmaster_get_option('payment', 'hipayprofessional-merchant-password', '');

					$parameters = new stdClass(); 
					$parameters->parameters = array(
						'wsLogin' => $username,
						'wsPassword' => $password,
						'transactionPublicId' => $transid
					);					

					$result = $client->getDetails($parameters);
					
					if ($result->getDetailsResult->code == "0" && $result->getDetailsResult->amount == $origAmount && $result->getDetailsResult->currency == $origCurrency && strtolower($result->getDetailsResult->transactionStatus) == "captured"){
	
			        	$tid = $idformerchant;
			        	$payment_info['transaction_id'] = $transid;
			        	$payment_info['amount'] = $origAmount;

			        	$tdata = tourmaster_get_booking_data(array('id'=>$tid), array('single'=>true));
						
						if( tourmaster_compare_price($tdata->total_price, $payment_info['amount']) ){
							$order_status = 'online-paid';
						} else {
							$order_status = 'pending';
						}

			        	tourmaster_update_booking_data( 
							array(
								'payment_info' => json_encode($payment_info),
								'payment_date' => current_time('mysql'),
								'order_status' => $order_status,
							),
							array('id' => $tid),
							array('%s', '%s', '%s'),
							array('%d')
						);

						tourmaster_mail_notification('payment-made-mail', $tid);
						tourmaster_mail_notification('admin-online-payment-made-mail', $tid);


					} 	

				} elseif ($status=="nok" || $operation == "cancellation" || $status == "cancel") {

		            $payment_info['error'] = $operation . " " . $status;
	            	$tid = $idformerchant;
					tourmaster_update_booking_data( 
						array(
							'payment_info' => json_encode($payment_info),
						),
						array('id' => $tid, 'payment_date' => '0000-00-00 00:00:00'),
						array('%s'),
						array('%d', '%s')
					);		            	
				}

	        exit;

		}

	} 
}